### PR TITLE
CauseEquals: new function, tests and regenerated documetnation

### PR DESCRIPTION
**Pull request by aurelijusb**
Useful when comparing constant error with returned from some deep function.

Useful in situations like:
```go
ErrNotFound := errors.New("Not found")
inSome := errors.Trace(ErrNotFound)
veryDeep := errors.Trace(inSome)
returned := errors.Trace(veryDeep)
if errors.CauseEquals(ErrNotFound, returned) {
    fmt.Println("Can return soft error 404")
} else {
    fmt.Println("Serious error, returning with code 500")
}
```

### CODE REVIEW

**Comment by howbazaar** [comment]
Isn't this the same as:

```go
if errors.Cause(returned) == ErrNotFound {
    fmt.Println("Can return soft error 404")
} else {
    fmt.Println("Serious error, returning with code 500")
}
```
?

File: functions_test.go [hunk 1]
```diff
@@ -303,3 +303,154 @@ func (*functionSuite) TestErrorStack(c *gc.C) {
...
+	level1 := errors.New("Level 1")
+	level2 := errors.Trace(level1)
+	level3 := errors.Trace(level2)
+	level4 := errors.Trace(level3)
+	level5 := errors.Trace(level4)
+	annotated6 := errors.Annotate(level5, "With annotation")
+	annotated7 := errors.Annotate(annotated6, "With another")
+	separate := errors.New("Separate")
+	var cycle error
+	cycle = errors.Trace(cycle)
```
Comment (howbazaar) [inline functions_test.go]: cycle here is still nil, and probably not what you expected.

File: functions.go [hunk 2]
```diff
@@ -328,3 +328,49 @@ func errorStack(err error) []string {
...
+			if maxDeep--; maxDeep <= 0 {
+				return false
+			}
+
+			if cause2, ok2 := error2.(*Err); ok2 {
+				error2 = cause2.Cause()
+			}
+		}
+		if cause1, ok1 := error1.(*Err); ok1 {
+			error1 = cause1.Cause()
```
Comment (howbazaar) [inline functions.go]: If I have 

```go
err := errors.New("some error")
```
then
```go
errors.Cause(err) == err // true
```

This would be making this for loop always exit due to maxDeep being hit.

**Comment by howbazaar** [comment]
I'm really struggling to find the real use for this. Can you explain the rationale a bit more?

File: README.md [hunk 3]
```diff
@@ -140,6 +147,26 @@ Cause is the usual way to diagnose errors that may have been wrapped by
 the other errors functions.
 
 
+## func CauseEquals
+``` go
+func CauseEquals(err1, err2 error) bool
+```
+CauseEquals tests, if this or any cause of this error is equal to another error, or it's cause
+
```
Comment (davecheney) [inline README.md]: Please move this unrelated documentation change to a new pull request. Thanks

File: functions.go [hunk 4]
```diff
@@ -328,3 +328,49 @@ func errorStack(err error) []string {
...
+// For example:
+//		ErrNotFound := errors.New("Not found")
+//		inSome := errors.Trace(ErrNotFound)
+//		veryDeep := errors.Trace(inSome)
+//		returned := errors.Trace(veryDeep)
+//		if errors.CauseEquals(ErrNotFound, returned) {
+//			fmt.Println("Can return soft error 404")
+//		} else {
+//			fmt.Println("Serious error, returning with code 500")
+//		}
```
Comment (davecheney) [inline functions.go]: Please move this text to an example test, 

func ExampleCausesEquals

File: functions.go [hunk 5]
```diff
@@ -328,3 +328,49 @@ func errorStack(err error) []string {
...
+func CauseEquals(err1, err2 error) bool {
+	var maxDeep = 1000
+	if err1 == err2 { // Same or nil == nil
+		return true
+	}
+	if (err1 == nil && err2 != nil) || (err2 == nil && err1 != nil) {
```
Comment (davecheney) [inline functions.go]: if err1 == nil || err2 == nil { .. } is sufficient, you've already asserted that err1 and err2 are not equal so therefore if err1 is nil, err2 is not and vice versa

File: functions.go [hunk 6]
```diff
@@ -328,3 +328,49 @@ func errorStack(err error) []string {
...
+//		ErrNotFound := errors.New("Not found")
+//		inSome := errors.Trace(ErrNotFound)
+//		veryDeep := errors.Trace(inSome)
+//		returned := errors.Trace(veryDeep)
+//		if errors.CauseEquals(ErrNotFound, returned) {
+//			fmt.Println("Can return soft error 404")
+//		} else {
+//			fmt.Println("Serious error, returning with code 500")
+//		}
+func CauseEquals(err1, err2 error) bool {
```
Comment (davecheney) [inline functions.go]: I have the same questions as @howbazaar but if this function goes ahead it should take the form

func CauseEquals(want, err error) bool {

that is, err should be recursively unpacked until it either cannot be unpacked further or matches `want`.

Comment (aurelijusb) [inline functions.go]: Hmm, for some reason was expecting that Cause will give only one level down, not the first cause (old habits from Java).
Now I realize, that with simple casting I could achieve the same.
```go
package main

import (
	"fmt"

	"github.com/juju/errors"
)

var ErrNotFound = errors.New("Resource not found")

func readElement() error  {
	return ErrNotFound;
}

func complexReadElement() error  {
	return errors.Trace(ErrNotFound);
}

func refactoredReadElement() error {
	if err := middlewareStuff(); err != nil {
		return errors.Trace(err)
	}
	return nil
}

func middlewareStuff() error {
	if err := lowLevelDatabaseRead(); err != nil {
		return errors.Trace(err)
	}
	return nil
}

func lowLevelDatabaseRead() error {
	return ErrNotFound
}

func annotatedRead() error {
	if err := middlewareStuff(); err != nil {
		return errors.Annotate(err, "Something")
	}
	return nil
}

func annotatedRead2() error {
	if err := annotatedRead(); err != nil {
		return errors.Annotate(err, "Something 2")
	}
	return nil
}

func main() {
	if readElement() == ErrNotFound {
		fmt.Println("Soft error 1")
	} else {
		fmt.Println("EXPECTING: Soft error 1")
	}

	if complexReadElement() == ErrNotFound {
		fmt.Println("Soft error 2")
	} else {
		fmt.Println("EXPECTING: Soft error 2")
	}

	// I do not want casting, but it is ok
	if complexReadElement().(*errors.Err).Cause() == ErrNotFound {
		fmt.Println("Soft error 3")
	} else {
		fmt.Println("EXPECTING: Soft error 3")
	}

	// Ok. Did not expect this :)
	if refactoredReadElement().(*errors.Err).Cause() == ErrNotFound {
		fmt.Println("Soft error 4")
	} else {
		fmt.Println("EXPECTING: Soft error 4")
	}

	// Ok. We could cast it and get what I expected.
	if refactoredReadElement().(*errors.Err).Cause() == middlewareStuff().(*errors.Err).Cause() {
		fmt.Println("Soft error 4")
	} else {
		fmt.Println("EXPECTING: Soft error 4")
	}

	// Ok. We could cast it and get what I expected.
	if annotatedRead2().(*errors.Err).Cause() == annotatedRead().(*errors.Err).Cause() {
		fmt.Println("Soft error 5")
	} else {
		fmt.Println("EXPECTING: Soft error 5")
	}
}
```
```
Soft error 1
EXPECTING: Soft error 2
Soft error 3
Soft error 4
Soft error 4
Soft error 5
```

**Comment by aurelijusb** [comment]
Alternatives are easy, so no need to include in main library.

File: functions.go (hunk 6 above)
Comment (howbazaar) [inline functions.go]: You don't need to do the casting, because that is what errors.Cause does.

Comment (aurelijusb) [inline functions.go]: Most of external libraries (which does not use `juju/errors`) returns error form standard `"errors"` package. So somewhere I will still need to do casting.

I mean:
```go
type error interface {
	Error() string
}
```
does not have function `Cause`

Comment (aurelijusb) [inline functions.go]: What about function to solve: **boilerplate of casting problem**?
```go
func CauseEquals(error1, error2 error) bool {
	var cause1 = error1
	var cause2 = error2
	if err1, ok := error1.(*errors.Err); ok && err1.Cause() != nil {
		cause1 = err1.Cause()
	}
	if err2, ok := error2.(*errors.Err); ok && err2.Cause() != nil {
		cause2 = err2.Cause()
	}
	return cause1 == cause2
}
```

Comment (aurelijusb) [inline functions.go]: So my [previous example](https://github.com/juju/errors/pull/22#discussion_r43275045) will look like:
```go
package main

import (
	"fmt"

	"github.com/juju/errors"
)

var ErrNotFound = errors.New("Resource not found")

func readElement() error  {
	return ErrNotFound;
}

func complexReadElement() error  {
	return errors.Trace(ErrNotFound);
}

func refactoredReadElement() error {
	if err := middlewareStuff(); err != nil {
		return errors.Trace(err)
	}
	return nil
}

func middlewareStuff() error {
	if err := lowLevelDatabaseRead(); err != nil {
		return errors.Trace(err)
	}
	return nil
}

func lowLevelDatabaseRead() error {
	return ErrNotFound
}

func annotatedRead() error {
	if err := middlewareStuff(); err != nil {
		return errors.Annotate(err, "Something")
	}
	return nil
}

func annotatedRead2() error {
	if err := annotatedRead(); err != nil {
		return errors.Annotate(err, "Something 2")
	}
	return nil
}

func main() {
	if CauseEquals(readElement(), ErrNotFound) {
		fmt.Println("Soft error 1")
	} else {
		fmt.Println("EXPECTING: Soft error 1")
	}

	if CauseEquals(complexReadElement(), ErrNotFound) {
		fmt.Println("Soft error 2")
	} else {
		fmt.Println("EXPECTING: Soft error 2")
	}

	// I do not want casting, but it is ok
	if CauseEquals(complexReadElement(), ErrNotFound) {
		fmt.Println("Soft error 3")
	} else {
		fmt.Println("EXPECTING: Soft error 3")
	}

	// Ok. Did not expect this :)
	if CauseEquals(refactoredReadElement(), ErrNotFound) {
		fmt.Println("Soft error 4")
	} else {
		fmt.Println("EXPECTING: Soft error 4")
	}

	// Ok. We could cast it and get what I expected.
	if CauseEquals(refactoredReadElement(), middlewareStuff()) {
		fmt.Println("Soft error 4")
	} else {
		fmt.Println("EXPECTING: Soft error 4")
	}

	// Ok. We could cast it and get what I expected.
	if CauseEquals(annotatedRead2(), annotatedRead()) {
		fmt.Println("Soft error 5")
	} else {
		fmt.Println("EXPECTING: Soft error 5")
	}
}

func CauseEquals(error1, error2 error) bool {
	var cause1 = error1
	var cause2 = error2
	if err1, ok := error1.(*errors.Err); ok && err1.Cause() != nil {
		cause1 = err1.Cause()
	}
	if err2, ok := error2.(*errors.Err); ok && err2.Cause() != nil {
		cause2 = err2.Cause()
	}
	return cause1 == cause2
}
```
```
Soft error 1
Soft error 2
Soft error 3
Soft error 4
Soft error 4
Soft error 5
```

Comment (howbazaar) [inline functions.go]: errors.Cause(err) is constructed so that it works with normal errors and errors from the juju/errors package.

https://github.com/juju/errors/blob/master/functions.go#L188

I tweaked your example to show what I mean:
```go
package main

import (
	"fmt"

	"github.com/juju/errors"
)

var ErrNotFound = errors.New("Resource not found")

func readElement() error {
	return ErrNotFound
}

func complexReadElement() error {
	return errors.Trace(ErrNotFound)
}

func refactoredReadElement() error {
	if err := middlewareStuff(); err != nil {
		return errors.Trace(err)
	}
	return nil
}

func middlewareStuff() error {
	if err := lowLevelDatabaseRead(); err != nil {
		return errors.Trace(err)
	}
	return nil
}

func lowLevelDatabaseRead() error {
	return ErrNotFound
}

func annotatedRead() error {
	if err := middlewareStuff(); err != nil {
		return errors.Annotate(err, "Something")
	}
	return nil
}

func annotatedRead2() error {
	if err := annotatedRead(); err != nil {
		return errors.Annotate(err, "Something 2")
	}
	return nil
}

func main() {
	if errors.Cause(readElement()) == ErrNotFound {
		fmt.Println("Soft error 1")
	} else {
		fmt.Println("EXPECTING: Soft error 1")
	}

	if errors.Cause(complexReadElement()) == ErrNotFound {
		fmt.Println("Soft error 2")
	} else {
		fmt.Println("EXPECTING: Soft error 2")
	}

	// I do not want casting, but it is ok
	if errors.Cause(complexReadElement()) == ErrNotFound {
		fmt.Println("Soft error 3")
	} else {
		fmt.Println("EXPECTING: Soft error 3")
	}

	// Ok. Did not expect this :)
	if errors.Cause(refactoredReadElement()) == ErrNotFound {
		fmt.Println("Soft error 4")
	} else {
		fmt.Println("EXPECTING: Soft error 4")
	}

	// Ok. We could cast it and get what I expected.
	if errors.Cause(refactoredReadElement()) == errors.Cause(middlewareStuff()) {
		fmt.Println("Soft error 5")
	} else {
		fmt.Println("EXPECTING: Soft error 5")
	}

	// Ok. We could cast it and get what I expected.
	if errors.Cause(annotatedRead2()) == errors.Cause(annotatedRead()) {
		fmt.Println("Soft error 6")
	} else {
		fmt.Println("EXPECTING: Soft error 6")
	}
}
```
```
Soft error 1
Soft error 2
Soft error 3
Soft error 4
Soft error 5
Soft error 6
```

Comment (aurelijusb) [inline functions.go]: Thanks for pointing this out :smile: 
```
func (e *Err) Cause() error
```
```
func Cause(err error) error
```
